flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaWriter.java

```diff
@@ -0,0 +1,379 @@
+
+    private final DeliveryGuarantee deliveryGuarantee;
+    private final Properties kafkaProducerConfig;
+    private final String transactionalIdPrefix;
+    private final KafkaRecordSerializationSchema<IN> recordSerializer;
+    private final Callback deliveryCallback;
+    private final AtomicLong pendingRecords = new AtomicLong();
+    private final KafkaRecordSerializationSchema.KafkaSinkContext kafkaSinkContext;
+    private final List<FlinkKafkaInternalProducer<byte[], byte[]>> producers = new ArrayList<>();
+    private final Map<String, KafkaMetricMutableWrapper> previouslyCreatedMetrics = new HashMap<>();
+    private final MetricGroup metricGroup;
+
+    private transient FlinkKafkaInternalProducer<byte[], byte[]> currentProducer;
+    private transient KafkaWriterState kafkaWriterState;
+    @Nullable private transient volatile Exception producerAsyncException;
+
+    /**
+     * Constructor creating a kafka writer.
+     *
+     * <p>It will throw a {@link RuntimeException} if {@link
+     * KafkaRecordSerializationSchema#open(SerializationSchema.InitializationContext)} fails.
+     *
+     * @param deliveryGuarantee the Sink's delivery guarantee
+     * @param kafkaProducerConfig the properties to configure the {@link FlinkKafkaInternalProducer}
+     * @param transactionalIdPrefix used to create the transactionalIds
+     * @param sinkInitContext context to provide information about the runtime environment
+     * @param recordSerializer serialize to transform the incoming records to {@link ProducerRecord}
+     * @param schemaContext context used to initialize the {@link KafkaRecordSerializationSchema}
+     * @param recoveredStates state from an previous execution which was covered
+     */
+    KafkaWriter(
+            DeliveryGuarantee deliveryGuarantee,
+            Properties kafkaProducerConfig,
+            String transactionalIdPrefix,
+            Sink.InitContext sinkInitContext,
+            KafkaRecordSerializationSchema<IN> recordSerializer,
+            SerializationSchema.InitializationContext schemaContext,
+            List<KafkaWriterState> recoveredStates) {
+        this.deliveryGuarantee = checkNotNull(deliveryGuarantee, "deliveryGuarantee");
+        this.kafkaProducerConfig = checkNotNull(kafkaProducerConfig, "kafkaProducerConfig");
+        this.transactionalIdPrefix = checkNotNull(transactionalIdPrefix, "transactionalIdPrefix");
+        this.recordSerializer = checkNotNull(recordSerializer, "recordSerializer");
+        try {
+            recordSerializer.open(schemaContext);
+        } catch (Exception e) {
+            throw new RuntimeException("Cannot initialize schema.", e);
+        }
+        this.deliveryCallback =
+                (metadata, exception) -> {
+                    if (exception != null && producerAsyncException == null) {
+                        producerAsyncException = exception;
+                    }
+                    acknowledgeMessage();
```

Comment:
We should now enqueue a mail to avoid the `AtomicLong` and `volatile`.

Reply:
Let's do this as a followup 👍